lib/utils/log/writer_finalizer.go

```diff
@@ -0,0 +1,49 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package log
+
+import (
+	"io"
+	"runtime"
+)
+
+// WriterFinalizer is a wrapper for the [io.WriteCloser] to automate resource cleanup.
+type WriterFinalizer struct {
+	writer io.WriteCloser
+}
```

Comment:
We only use this with concrete types, we can avoid the virtualization cost I think?

```suggestion
type WriterFinalizer[T io.WriteCloser] struct {
	writer T
}
```